File: LINQ-Task3/Generic/src/main/java/org/example/RandomizedList.java
package org.example;

import java.util.ArrayList;
import java.util.Random;

public class RandomizedList<T> {
    private final ArrayList<T> items;
    private final Random random;

    public RandomizedList() {
        this.items = new ArrayList<>();
        this.random = new Random();
    }

    public void add(T element) {
        if (random.nextBoolean()) {
            items.add(0, element);
        } else {
            items.add(element);
        }
    }

    public T get(int index) {
        if (items.isEmpty() || index >= items.size()) {
            return null;
        }
        return items.get(random.nextInt(index + 1));
    }

    public boolean isEmpty() {
        return items.isEmpty();
    }
}

File: 03-LINQ/GoldSavings.App/Model/RandomizedList.cs
namespace GoldSavings.App.Model;

public class RandomizedList<T>
{
    private List<T> _list = new List<T>();
    private Random _random = new Random();
    
    public void Add(T element)
    {
        if (_random.Next(0, 2) == 0)
        {
            _list.Insert(0, element);
        }
        else
        {
            _list.Add(element);
        }
    }
    
    public T Get(int index)
    {
        if (_list.Count == 0)
        {
            throw new Exception("List is empty.");
        }
        
        return _list[_random.Next(0, index)];
    }
    
    public bool IsEmpty()
    {
        return _list.Count == 0;
    }
}